src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,362 @@
+            {
+                "area",
+                "base",
+                "br",
+                "col",
+                "command",
+                "embed",
+                "hr",
+                "img",
+                "input",
+                "keygen",
+                "link",
+                "meta",
+                "param",
+                "source",
+                "track",
+                "wbr"
+            };
+
+            private readonly Stack<TagBlockTracker> _startTagTracker;
+
+            public Rewriter()
+            {
+                _startTagTracker = new Stack<TagBlockTracker>();
+            }
+
+            private TagBlockTracker CurrentTracker => _startTagTracker.Count > 0 ? _startTagTracker.Peek() : null;
+
+            private string CurrentStartTagName => CurrentTracker?.TagName;
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
+
+                return base.Visit(node);
+            }
+
+            private SyntaxNode RewriteNode(SyntaxNode node)
+            {
+                if (node.IsToken ||
+                    node is MarkupElementSyntax ||
+                    node.Parent is MarkupElementSyntax)
+                {
+                    // This is either a token or we have already rewritten this node.
+                    return node;
+                }
+
+                var currentIndex = 0;
+                _startTagTracker.Clear();
+                var children = node.ChildNodes();
+                while (currentIndex < children.Count)
+                {
+                    var child = (RazorSyntaxNode)children[currentIndex];
+                    if (!(child is MarkupTagBlockSyntax tagBlock) ||
+                        tagBlock.Parent is MarkupElementSyntax)
+                    {
+                        // Not a node that we want to rewrite.
+                        currentIndex++;
+                        continue;
+                    }
+
+                    var tagName = tagBlock.GetTagName();
+                    if (string.IsNullOrWhiteSpace(tagName) ||
+                        IsVoidElement(tagBlock) ||
+                        IsSelfClosing(tagBlock))
+                    {
+                        // Don't want to track incomplete, void or self-closing tags.
+                        // Simply wrap it in a block with no body or end tag.
+                        node = RewriteNodeCore(node, startTag: tagBlock, tagChildren: Enumerable.Empty<RazorSyntaxNode>(), endTag: null);
+
+                        // Since we rewrote 'node', it's children are different. Update our collection.
+                        children = node.ChildNodes();
+                    }
+                    else if (IsEndTag(tagBlock))
+                    {
+                        if (!string.IsNullOrEmpty(CurrentStartTagName) &&
```

Comment:
You can remove this check. Next line should suffice since you already verify `tagName` has a value